src/models/builders/services/vercel-service.ts

```diff
@@ -0,0 +1,109 @@
+import { CannotWriteFile, VercelProjectError, VercelLoadConfigError} from "./errors/error";
+import { mkdirSync, readFileSync, writeFileSync, existsSync} from "fs";
+import { dirname, join, relative, resolve } from "path";
+import glob from "fast-glob";
+import { tmpdir } from "os";
+
+import util = require('node:util');
+// eslint-disable-next-line @typescript-eslint/no-var-requires
+const exec = util.promisify(require('node:child_process').exec);
+
+export class VercelService {
+    tmpFunctionsDir: string = join(tmpdir(), Math.random().toString(36).slice(2));
+
+    createVercelProjectConfig() {
+        console.log("Creating project config file ...");
+    
+        try {
+            const projectConfigDir = '.vercel';
+            const projectConfigFilePath = `${projectConfigDir}/project.json`;
+    
+            if (!existsSync(projectConfigFilePath)) {
+                if (!existsSync(projectConfigDir)) {
+                    mkdirSync(projectConfigDir);
+                }
+    
+                writeFileSync(projectConfigFilePath,'{"projectId":"_","orgId":"_","settings":{}}');
+            }
+        } catch (error:any) {
+            throw new CannotWriteFile(error.message);
+        }
+    }
+
+    async runVercelBuild() {
+        // https://vercel.com/docs/build-output-api/v3
+        console.log("Running initial build ...");
+
+        try {
+            await exec('npx --yes vercel@28.16.11 build --prod');
+        } catch (error:any) {
+            throw new VercelProjectError(error.message);
+        }
+    }
+
+    loadVercelConfigs(): any {
+        console.log("Loading configs ...")
+
+        try {
+            const fileContent = readFileSync(".vercel/output/config.json", "utf8");
+            const config = JSON.parse(fileContent);
+
+            return config;
+        } catch (error:any) {
+            throw new VercelLoadConfigError(error.message);
+        }
+    }
+
+    // function to walk in builded functions dir, detect invalid functions and adapt content
+    adapt() {
+        try{
+            // eslint-disable-next-line prefer-const
+            let functionsMap: Map<string, string> = new Map();
+            const vcConfigPaths: Array<string> = glob.sync(".vercel/output/functions/**/.vc-config.json");
+            const vcConfigObjects:Array<any> = vcConfigPaths.map(file => {
+                return {
+                    path: file,
+                    content: JSON.parse(readFileSync(file, "utf8")),
+                }
+            });
+            const validVcObjects:Array<any> = vcConfigObjects.filter(vcConfig =>  this.isVcConfigValid(vcConfig.content));
```

Comment:
Maybe we can use `reduce` or `foreach` to avoid doing the filter twice.

Reply:
lets keep this refactor to another pull request as we talked about